proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java

```diff
@@ -89,18 +100,27 @@ public int getEncodeVersion()
 
     public void write(Object packet)
     {
+        // ensure netty context to for less context schedules
+        // by default we are mostly in netty context anyway, but plugins can use ProxiedPlayer.unsafe().sendPacket()
+        // in non netty context
+        if ( !ch.eventLoop().inEventLoop() )
+        {
+            ch.eventLoop().execute( () -> write( packet ) );
+            return;
+        }
+
         if ( !closed )
         {
             DefinedPacket defined = null;
             if ( packet instanceof PacketWrapper )
             {
                 PacketWrapper wrapper = (PacketWrapper) packet;
                 wrapper.setReleased( true );
-                ch.writeAndFlush( wrapper.buf, ch.voidPromise() );
+                ch.write( wrapper.buf, ch.voidPromise() );
```

Comment:
Instead of this construction, wouldn't it be simpler if we add a write0 method in ChannelWrapper which has the logic of signalFlush?

Reply:
why?

signalFlush is called guaranteed at the end of the method. Only if the encoder protocol is changed it will forceFlush and return.